### PR TITLE
Add e2e test for mounting different paths on same volume on same node

### PR DESCRIPTION
**Pull request by wongma7**
**Is this a bug fix or adding new feature?** fix https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/105

**What is this PR about? / Why do we need it?** #100 was fixed but we need an e2e test for it.

**What testing is done?** 
Testing on my EKS cluster:
```
------------------------------
[efs-csi] EFS CSI [Driver: efs.csi.aws.com] 
  should mount different paths on same volume on same node
  /Users/mattwon/go/src/github.com/kubernetes-sigs/aws-efs-csi-driver/test/e2e/e2e.go:182
[BeforeEach] [efs-csi] EFS CSI
  /Users/mattwon/go/pkg/mod/k8s.io/kubernetes@v1.17.6/test/e2e/framework/framework.go:151
STEP: Creating a kubernetes client
Jun  1 16:00:48.241: INFO: >>> kubeConfig: /Users/mattwon/.kube/config
STEP: Building a namespace api object, basename efs
Jun  1 16:00:48.320: INFO: Found PodSecurityPolicies; assuming PodSecurityPolicy is enabled.
Jun  1 16:00:48.362: INFO: Found ClusterRoles; assuming RBAC is enabled.
STEP: Binding the e2e-test-privileged-psp PodSecurityPolicy to the default service account in efs-6908
STEP: Waiting for a default service account to be provisioned in namespace
[It] should mount different paths on same volume on same node
  /Users/mattwon/go/src/github.com/kubernetes-sigs/aws-efs-csi-driver/test/e2e/e2e.go:182
STEP: Creating efs pvc & pv with no subpath
STEP: Creating pod to make subpaths /a and /b
Jun  1 16:00:48.562: INFO: Waiting up to 5m0s for pod "pvc-tester-frqh5" in namespace "efs-6908" to be "success or failure"
Jun  1 16:00:48.577: INFO: Pod "pvc-tester-frqh5": Phase="Pending", Reason="", readiness=false. Elapsed: 15.387249ms
Jun  1 16:00:50.596: INFO: Pod "pvc-tester-frqh5": Phase="Pending", Reason="", readiness=false. Elapsed: 2.03421632s
Jun  1 16:00:52.612: INFO: Pod "pvc-tester-frqh5": Phase="Pending", Reason="", readiness=false. Elapsed: 4.050449793s
Jun  1 16:00:54.628: INFO: Pod "pvc-tester-frqh5": Phase="Pending", Reason="", readiness=false. Elapsed: 6.066294845s
Jun  1 16:00:56.645: INFO: Pod "pvc-tester-frqh5": Phase="Pending", Reason="", readiness=false. Elapsed: 8.082881893s
Jun  1 16:00:58.662: INFO: Pod "pvc-tester-frqh5": Phase="Pending", Reason="", readiness=false. Elapsed: 10.1000315s
Jun  1 16:01:00.683: INFO: Pod "pvc-tester-frqh5": Phase="Pending", Reason="", readiness=false. Elapsed: 12.121066843s
Jun  1 16:01:02.699: INFO: Pod "pvc-tester-frqh5": Phase="Pending", Reason="", readiness=false. Elapsed: 14.13670026s
Jun  1 16:01:04.715: INFO: Pod "pvc-tester-frqh5": Phase="Pending", Reason="", readiness=false. Elapsed: 16.153338265s
Jun  1 16:01:06.734: INFO: Pod "pvc-tester-frqh5": Phase="Pending", Reason="", readiness=false. Elapsed: 18.171695624s
Jun  1 16:01:08.753: INFO: Pod "pvc-tester-frqh5": Phase="Pending", Reason="", readiness=false. Elapsed: 20.191351553s
Jun  1 16:01:10.770: INFO: Pod "pvc-tester-frqh5": Phase="Pending", Reason="", readiness=false. Elapsed: 22.208355608s
Jun  1 16:01:12.787: INFO: Pod "pvc-tester-frqh5": Phase="Pending", Reason="", readiness=false. Elapsed: 24.225318444s
Jun  1 16:01:14.805: INFO: Pod "pvc-tester-frqh5": Phase="Pending", Reason="", readiness=false. Elapsed: 26.243291824s
Jun  1 16:01:16.824: INFO: Pod "pvc-tester-frqh5": Phase="Pending", Reason="", readiness=false. Elapsed: 28.262323831s
Jun  1 16:01:18.841: INFO: Pod "pvc-tester-frqh5": Phase="Pending", Reason="", readiness=false. Elapsed: 30.278845233s
Jun  1 16:01:20.858: INFO: Pod "pvc-tester-frqh5": Phase="Pending", Reason="", readiness=false. Elapsed: 32.295765653s
Jun  1 16:01:22.873: INFO: Pod "pvc-tester-frqh5": Phase="Pending", Reason="", readiness=false. Elapsed: 34.311190886s
Jun  1 16:01:24.891: INFO: Pod "pvc-tester-frqh5": Phase="Pending", Reason="", readiness=false. Elapsed: 36.328770405s
Jun  1 16:01:26.906: INFO: Pod "pvc-tester-frqh5": Phase="Pending", Reason="", readiness=false. Elapsed: 38.343785078s
Jun  1 16:01:28.922: INFO: Pod "pvc-tester-frqh5": Phase="Pending", Reason="", readiness=false. Elapsed: 40.360238146s
Jun  1 16:01:30.943: INFO: Pod "pvc-tester-frqh5": Phase="Succeeded", Reason="", readiness=false. Elapsed: 42.38106425s
STEP: Saw pod success
Jun  1 16:01:30.943: INFO: Pod "pvc-tester-frqh5" satisfied condition "success or failure"
STEP: Creating efs pvc & pv with subpath /a
STEP: Creating efs pvc & pv with subpath /b
STEP: Creating pod on node "ip-192-168-83-235.us-west-2.compute.internal" to mount subpaths /a and /b
[AfterEach] [efs-csi] EFS CSI
  /Users/mattwon/go/pkg/mod/k8s.io/kubernetes@v1.17.6/test/e2e/framework/framework.go:152
Jun  1 16:02:13.136: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
STEP: Destroying namespace "efs-6908" for this suite.

• [SLOW TEST:84.927 seconds]
[efs-csi] EFS CSI
/Users/mattwon/go/src/github.com/kubernetes-sigs/aws-efs-csi-driver/test/e2e/e2e.go:173
  [Driver: efs.csi.aws.com]
  /Users/mattwon/go/src/github.com/kubernetes-sigs/aws-efs-csi-driver/test/e2e/e2e.go:181
    should mount different paths on same volume on same node
    /Users/mattwon/go/src/github.com/kubernetes-sigs/aws-efs-csi-driver/test/e2e/e2e.go:182
------------------------------
```
Testing on CI's kops cluster (https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_aws-efs-csi-driver/173/pull-aws-efs-csi-driver-e2e/1268250579340103680):
```
 [efs-csi] EFS CSI [Driver: efs.csi.aws.com] 
  should mount different paths on same volume on same node
  /home/prow/go/src/github.com/kubernetes-sigs/aws-efs-csi-driver/test/e2e/e2e.go:182
[BeforeEach] [efs-csi] EFS CSI
  /home/prow/go/pkg/mod/k8s.io/kubernetes@v1.17.6/test/e2e/framework/framework.go:151
STEP: Creating a kubernetes client
Jun  3 19:07:41.652: INFO: >>> kubeConfig: /root/.kube/config
STEP: Building a namespace api object, basename efs
STEP: Waiting for a default service account to be provisioned in namespace
[It] should mount different paths on same volume on same node
  /home/prow/go/src/github.com/kubernetes-sigs/aws-efs-csi-driver/test/e2e/e2e.go:182
STEP: Creating efs pvc & pv with no subpath
STEP: Creating pod to make subpaths /a and /b
Jun  3 19:07:42.041: INFO: Waiting up to 5m0s for pod "pvc-tester-lndpk" in namespace "efs-1888" to be "success or failure"
Jun  3 19:07:42.096: INFO: Pod "pvc-tester-lndpk": Phase="Pending", Reason="", readiness=false. Elapsed: 54.842217ms
Jun  3 19:07:44.154: INFO: Pod "pvc-tester-lndpk": Phase="Pending", Reason="", readiness=false. Elapsed: 2.112854684s
Jun  3 19:07:46.209: INFO: Pod "pvc-tester-lndpk": Phase="Pending", Reason="", readiness=false. Elapsed: 4.167987328s
Jun  3 19:07:48.264: INFO: Pod "pvc-tester-lndpk": Phase="Pending", Reason="", readiness=false. Elapsed: 6.222731898s
Jun  3 19:07:50.319: INFO: Pod "pvc-tester-lndpk": Phase="Pending", Reason="", readiness=false. Elapsed: 8.277866927s
Jun  3 19:07:52.375: INFO: Pod "pvc-tester-lndpk": Phase="Succeeded", Reason="", readiness=false. Elapsed: 10.333033294s
STEP: Saw pod success
Jun  3 19:07:52.375: INFO: Pod "pvc-tester-lndpk" satisfied condition "success or failure"
STEP: Creating efs pvc & pv with subpath /a
STEP: Creating efs pvc & pv with subpath /b
STEP: Creating pod on node "ip-172-20-114-53.us-west-2.compute.internal" to mount subpaths /a and /b
[AfterEach] [efs-csi] EFS CSI
  /home/prow/go/pkg/mod/k8s.io/kubernetes@v1.17.6/test/e2e/framework/framework.go:152
Jun  3 19:08:12.988: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
STEP: Destroying namespace "efs-1888" for this suite. 
```

### CODE REVIEW

**Comment by wongma7** [comment]
/test pull-aws-efs-csi-driver-e2e

https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_aws-efs-csi-driver/173/pull-aws-efs-csi-driver-e2e/1268220967159599105 looks like an instance of https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/141. Saw similar 'mount.nfs4...blocked for more than 120 seconds.' messages in dmesg with no mounts of fs-286c882d on node ip-172-20-119-122.us-west-2.compute.internal ever succeeding

**Comment by wongma7** [comment]
/assign @leakingtapan 
PTAL if you get a chance, this is a test to verify the fix you implemented for the path mounting issue. (https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/105). THX

File: test/e2e/e2e.go [hunk 1]
```diff
@@ -172,4 +175,108 @@ var _ = ginkgo.Describe("[efs-csi] EFS CSI", func() {
...
+			pvcA, pvA, err := createEFSPVCPV(f.ClientSet, f.Namespace.Name, f.Namespace.Name+"-a", "/a")
+			framework.ExpectNoError(err, "creating efs pvc & pv with subpath /a")
+			defer func() { _ = f.ClientSet.CoreV1().PersistentVolumes().Delete(pvA.Name, &metav1.DeleteOptions{}) }()
+
+			ginkgo.By(fmt.Sprintf("Creating efs pvc & pv with subpath /b"))
+			pvcB, pvB, err := createEFSPVCPV(f.ClientSet, f.Namespace.Name, f.Namespace.Name+"-b", "/b")
+			framework.ExpectNoError(err, "creating efs pvc & pv with subpath /b")
+			defer func() { _ = f.ClientSet.CoreV1().PersistentVolumes().Delete(pvB.Name, &metav1.DeleteOptions{}) }()
+
+			node, err := e2enode.GetRandomReadySchedulableNode(f.ClientSet)
```
Comment (wongma7) [inline test/e2e/e2e.go:204]: oops, this isn't needed for this test since it's just creating 1 pod.

File: test/e2e/e2e.go [hunk 2]
```diff
@@ -172,4 +174,105 @@ var _ = ginkgo.Describe("[efs-csi] EFS CSI", func() {
...
+
+	ginkgo.Context(testsuites.GetDriverNameWithFeatureTags(driver), func() {
+		ginkgo.It("should mount different paths on same volume on same node", func() {
+			ginkgo.By(fmt.Sprintf("Creating efs pvc & pv with no subpath"))
+			pvcRoot, pvRoot, err := createEFSPVCPV(f.ClientSet, f.Namespace.Name, f.Namespace.Name+"-root", "/")
+			framework.ExpectNoError(err, "creating efs pvc & pv with no subpath")
+			defer func() { _ = f.ClientSet.CoreV1().PersistentVolumes().Delete(pvRoot.Name, &metav1.DeleteOptions{}) }()
+
+			ginkgo.By(fmt.Sprintf("Creating pod to make subpaths /a and /b"))
+			pod := e2epod.MakePod(f.Namespace.Name, nil, []*v1.PersistentVolumeClaim{pvcRoot}, false, "mkdir /mnt/volume1/a && mkdir /mnt/volume1/b")
```
Comment (jqmichael) [inline test/e2e/e2e.go:188]: Was "/mnt/volume1" defined in the pod spec?

```
		volumeMounts[index] = v1.VolumeMount{Name: volumename, MountPath: "/mnt/" + volumename}
```
https://github.com/kubernetes/kubernetes/blob/7c161d6a5ff000de2147a33fd62225f398d76ae2/test/e2e/framework/pod/create.go#L190

File: test/e2e/e2e.go [hunk 3]
```diff
@@ -172,4 +174,105 @@ var _ = ginkgo.Describe("[efs-csi] EFS CSI", func() {
...
+			pod, err = f.ClientSet.CoreV1().Pods(f.Namespace.Name).Create(pod)
+			framework.ExpectNoError(err, "creating pod")
+			framework.ExpectNoError(e2epod.WaitForPodSuccessInNamespace(f.ClientSet, pod.Name, f.Namespace.Name), "waiting for pod success")
+
+			ginkgo.By(fmt.Sprintf("Creating efs pvc & pv with subpath /a"))
+			pvcA, pvA, err := createEFSPVCPV(f.ClientSet, f.Namespace.Name, f.Namespace.Name+"-a", "/a")
```
Comment (jqmichael) [inline test/e2e/e2e.go:194]: Wish Golang support keywords argument like Python does. It's not easy to tell what those parameters really represent.

File: test/e2e/e2e.go (hunk 2 above)
Comment (jqmichael) [inline test/e2e/e2e.go:188]: The interface of `MakePod`

```
func MakePod(ns string, nodeSelector map[string]string, pvclaims []*v1.PersistentVolumeClaim, isPrivileged bool, command string) *v1.Pod {
```

https://github.com/kubernetes/kubernetes/blob/7c161d6a5ff000de2147a33fd62225f398d76ae2/test/e2e/framework/pod/create.go#L158

File: test/e2e/e2e.go [hunk 4]
```diff
@@ -172,4 +174,105 @@ var _ = ginkgo.Describe("[efs-csi] EFS CSI", func() {
...
+			ginkgo.By(fmt.Sprintf("Creating efs pvc & pv with subpath /b"))
+			pvcB, pvB, err := createEFSPVCPV(f.ClientSet, f.Namespace.Name, f.Namespace.Name+"-b", "/b")
+			framework.ExpectNoError(err, "creating efs pvc & pv with subpath /b")
+			defer func() { _ = f.ClientSet.CoreV1().PersistentVolumes().Delete(pvB.Name, &metav1.DeleteOptions{}) }()
+
+			ginkgo.By("Creating pod on to mount subpaths /a and /b")
+			pod = e2epod.MakePod(f.Namespace.Name, nil, []*v1.PersistentVolumeClaim{pvcA, pvcB}, false, "")
+			pod, err = f.ClientSet.CoreV1().Pods(f.Namespace.Name).Create(pod)
+			framework.ExpectNoError(err, "creating pod")
+			framework.ExpectNoError(e2epod.WaitForPodNameRunningInNamespace(f.ClientSet, pod.Name, f.Namespace.Name), "waiting for pod running")
```
Comment (jqmichael) [inline test/e2e/e2e.go:207]: Is checking the mount is accessible necessary, i.e. `kubectl exec ... ls /mnt/volume1`?

File: test/e2e/e2e.go [hunk 5]
```diff
@@ -172,4 +174,105 @@ var _ = ginkgo.Describe("[efs-csi] EFS CSI", func() {
...
 	})
+
+	f := framework.NewDefaultFramework("efs")
+
+	ginkgo.Context(testsuites.GetDriverNameWithFeatureTags(driver), func() {
+		ginkgo.It("should mount different paths on same volume on same node", func() {
+			ginkgo.By(fmt.Sprintf("Creating efs pvc & pv with no subpath"))
+			pvcRoot, pvRoot, err := createEFSPVCPV(f.ClientSet, f.Namespace.Name, f.Namespace.Name+"-root", "/")
+			framework.ExpectNoError(err, "creating efs pvc & pv with no subpath")
+			defer func() { _ = f.ClientSet.CoreV1().PersistentVolumes().Delete(pvRoot.Name, &metav1.DeleteOptions{}) }()
```
Comment (jqmichael) [inline test/e2e/e2e.go:185]: Are PVC and Pods getting cleaned up automatically?

Comment (wongma7) [reply test/e2e/e2e.go:185]: yes, since namespaces gets cleaned up automatically

File: test/e2e/e2e.go (hunk 3 above)
Comment (wongma7) [reply test/e2e/e2e.go:194]: agree lol. a pattern I have seen is to add /* <parameter name> */ after every arg but thats gets tiring and verbose fast. IDE can help here.

File: test/e2e/e2e.go (hunk 4 above)
Comment (wongma7) [reply test/e2e/e2e.go:207]: no, the pod from the initial step "ginkgo.By(fmt.Sprintf("Creating pod to make subpaths /a and /b"))" would have failed if somehow mnt is not accessible

**Pull request merged**